linera-execution/src/evm/revm.rs

```diff
@@ -1108,33 +1016,90 @@ impl<Runtime: ContractRuntime> CallInterceptorContract<Runtime> {
     fn call_or_fail(
         &mut self,
         context: &mut Ctx<'_, Runtime>,
         inputs: &mut CallInputs,
     ) -> Result<Option<CallOutcome>, ExecutionError> {
-        // Every call to a contract passes by this function.
-        // Three kinds:
-        // --- Call to the PRECOMPILE smart contract.
-        // --- Call to the EVM smart contract itself
-        // --- Call to other EVM smart contract
         if self.precompile_addresses.contains(&inputs.target_address)
             || inputs.target_address == self.contract_address
         {
             // Precompile calls are handled by the precompile code.
             // The EVM smart contract is being called
             return Ok(None);
         }
+        // Handling the balances.
+        if let CallValue::Transfer(value) = inputs.value {
+            let source: AccountOwner = inputs.caller.into();
+            let owner: AccountOwner = inputs.bytecode_address.into();
+            if value != U256::ZERO {
+                // In Linera, only non-zero transfers matter
+                {
+                    let mut runtime = context
+                        .db()
+                        .0
+                        .runtime
+                        .lock()
+                        .expect("The lock should be possible");
+                    let amount = Amount::try_from(value).map_err(EvmExecutionError::from)?;
+                    let chain_id = runtime.chain_id()?;
+                    let destination = Account { chain_id, owner };
+                    runtime.transfer(source, destination, amount)?;
+                }
+                revm_transfer(context, inputs.caller, inputs.target_address, value)?;
+            }
+        }
         // Other smart contracts calls are handled by the runtime
         let target = address_to_user_application_id(inputs.target_address);
-        let argument = get_call_argument(context, inputs);
-        let authenticated = true;
-        let result = {
-            let mut runtime = self.db.runtime.lock().unwrap();
-            runtime.try_call_application(authenticated, target, argument)?
+        let (argument, n_input) = get_call_contract_argument(context, inputs)?;
+        let result = if n_input > 0 {
+            // The input is non-trivial, we assume that we are calling a contract.
+            //
+            // The correct behavior is the following:
+            // * If a contract exists and input is empty, then the fallback function is called
+            //   if existing. Otherwise failure.
+            // * If not, then it is a user account and no execution occurs.
+            //
+            // So, the correct way is instead to test the existence of the application
+            // given the application_id. So, we would need following function in BaseRuntime:
+            //  fn is_existing_application(&mut self, application_id: ApplicationId)
+            //       -> Result<bool, ExecutionError>;
```

Comment:
Is this comment meant as a "TODO"? (If so, it needs a GitHub issue.)

I'm still not sure what the right behavior is, though. A runtime call that is purely local to the chain can only check whether an application has already been used _on that chain_, not whether it exists in the system. The latter could only be done as an oracle call.

Reply:
Now merged and used in the code.